src/sdi-progress-dock.c

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (C) 2020-2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "sdi-progress-dock.h"
+#include "com.canonical.Unity.LauncherEntry.h"
+#include <glib/gi18n.h>
+#include <snapd-glib/snapd-glib.h>
+
+/**
+ * This class manages the progress bars in the dock for each snap
+ * being updated.
+ */
+
+struct _SdiProgressDock {
+  GObject parent_instance;
+
+  GApplication *application;
+  UnityComCanonicalUnityLauncherEntry *unity_manager;
+};
+
+G_DEFINE_TYPE(SdiProgressDock, sdi_progress_dock, G_TYPE_OBJECT)
+
+/**
+ * This callback should be connected to the `update-progress` signal from a
+ * #sdi_refresh_monitor object. It will receive the total number of tasks and
+ * how many have been done, and if the task has been completed, and with that
+ * will update the progress bars in the dock.
+ */
+void sdi_progress_dock_update_progress(SdiProgressDock *self, gchar *snap_name,
+                                       GStrv desktop_files,
+                                       gchar *task_description,
+                                       guint done_tasks, guint total_tasks,
+                                       gboolean task_done, gpointer data) {
```

Comment:
Yes, I know that it works, but... I really dislike doing that. Yes, *in theory*, in C, the caller is responsible for cleaning the stack, and since the parameters are stacked "in reverse", any extra parameter just won't be accessed, so they shouldn't be a problem... unless the calling convention says a different thing, like in the case of https://learn.microsoft.com/en-us/cpp/cpp/stdcall?view=msvc-170 (yes, I know that it is a little bit elaborated).

Anyway... the other cases do the same, so... for consistency...?

Reply:
So, my point is that those functions are publicly accessible methods of the dock class, so they should be something that is potentially callable also directly, and not considered as signal callbacks.

If you want signal callbacks, then you can use local functions that adapt the callback arguments to the function that is being invoked.

Now, since in this case the signature match, we can just rely on `G_CALLBACK` doing its thing.

If you instead prefer to have a callback that has the whole signature, then use local functions to call them, but to me it looks just an unneeded repetition of code.

Reply:
Well, after checking the GLib code, I see that the trick of relying on G_CALLBACK to remove the `data` parameter is used there, so... Ok.

Reply:
Done on all cases, not only on this one.